File: frontend/src/api/attackPath.ts
/**
 * API client for Attack Path Analysis
 */
import { apiClient } from './index';
import type { AxiosResponse } from 'axios';

// Enums
export enum AttackPathType {
  DIRECT = 'Direct',
  MULTI_STEP = 'Multi-Step',
  LATERAL = 'Lateral',
  PRIVILEGE_ESCALATION = 'Privilege Escalation'
}

export enum AttackStepType {
  INITIAL_ACCESS = 'Initial Access',
  EXECUTION = 'Execution',
  PERSISTENCE = 'Persistence',
  PRIVILEGE_ESCALATION = 'Privilege Escalation',
  DEFENSE_EVASION = 'Defense Evasion',
  CREDENTIAL_ACCESS = 'Credential Access',
  DISCOVERY = 'Discovery',
  LATERAL_MOVEMENT = 'Lateral Movement',
  COLLECTION = 'Collection',
  EXFILTRATION = 'Exfiltration',
  COMMAND_AND_CONTROL = 'Command and Control',
  IMPACT = 'Impact'
}

export enum AttackComplexity {
  LOW = 'Low',
  MEDIUM = 'Medium',
  HIGH = 'High'
}

// Interfaces
export interface ImpactScores {
  confidentiality: number;
  integrity: number;
  availability: number;
}

// Fix for AttackPathAssumption import issue
export class AttackPathAssumption {
  assumption_id: string;
  description: string;
  type: string;

  constructor(assumption_id: string, description: string, type: string) {
    this.assumption_id = assumption_id;
    this.description = description;
    this.type = type;
  }
}

// Original interface kept for reference
export interface IAttackPathAssumption {
  assumption_id: string;
  description: string;
  type: string;
}

// Fix for AttackPathConstraint import issue
export class AttackPathConstraint {
  constraint_id: string;
  description: string;
  type: string;

  constructor(constraint_id: string, description: string, type: string) {
    this.constraint_id = constraint_id;
    this.description = description;
    this.type = type;
  }
}

// Original interface kept for reference
export interface IAttackPathConstraint {
  constraint_id: string;
  description: string;
  type: string;
}

// Fix for ThreatScenario import issue
export class ThreatScenario {
  scenario_id: string;
  name: string;
  description: string;
  threat_type: string;
  likelihood: number;

  constructor(scenario_id: string, name: string, description: string, threat_type: string, likelihood: number) {
    this.scenario_id = scenario_id;
    this.name = name;
    this.description = description;
    this.threat_type = threat_type;
    this.likelihood = likelihood;
  }
}

// Original interface kept for reference
export interface IThreatScenario {
  scenario_id: string;
  name: string;
  description: string;
  threat_type: string;
  likelihood: number;
}

export interface AttackStep {
  step_id: string;
  component_id: string;
  step_type: AttackStepType;
  description: string;
  prerequisites?: string;
  vulnerability_ids?: string[];
  threat_ids?: string[];
  order: number;
  created_at: string;
}

export interface AttackPath {
  path_id: string;
  name: string;
  description: string;
  path_type: AttackPathType;
  complexity: AttackComplexity;
  entry_point_id: string;
  target_id: string;
  success_likelihood: number;
  impact: ImpactScores;
  risk_score: number;
  analysis_id: string;
  scope_id: string | null;
  created_at: string;
  updated_at: string;
  steps: AttackStep[];
}

export interface AttackPathList {
  paths: AttackPath[];
  total: number;
}

export interface AttackChainPath {
  path_id: string;
  chain_id: string;
  order: number;
}

export interface AttackChain {
  chain_id: string;
  name: string;
  description: string;
  entry_point_id: string;
  final_target_id: string;
  total_steps: number;
  complexity: AttackComplexity;
  success_likelihood: number;
  impact: ImpactScores;
  risk_score: number;
  analysis_id: string;
  scope_id: string | null;
  created_at: string;
  updated_at: string;
  paths: AttackPath[];
}

export interface AttackChainList {
  chains: AttackChain[];
  total: number;
}

export interface ComponentSummary {
  component_id: string;
  name: string;
  type: string;
  safety_level?: string;
  trust_zone?: string;
}

// Fix for AttackPathRequest import issue
export class AttackPathRequest {
  primary_component_id: string;
  component_ids: string[];
  entry_point_ids?: string[];
  target_ids?: string[];
  include_chains: boolean;
  max_depth?: number;
  assumptions?: AttackPathAssumption[];
  constraints?: AttackPathConstraint[];
  threat_scenarios?: ThreatScenario[];
  vulnerability_ids?: string[];

  constructor(args: {
    primary_component_id: string;
    component_ids: string[];
    entry_point_ids?: string[];
    target_ids?: string[];
    include_chains: boolean;
    max_depth?: number;
    assumptions?: AttackPathAssumption[];
    constraints?: AttackPathConstraint[];
    threat_scenarios?: ThreatScenario[];
    vulnerability_ids?: string[];
  }) {
    this.primary_component_id = args.primary_component_id;
    this.component_ids = args.component_ids;
    this.entry_point_ids = args.entry_point_ids;
    this.target_ids = args.target_ids;
    this.include_chains = args.include_chains;
    this.max_depth = args.max_depth;
    this.assumptions = args.assumptions;
    this.constraints = args.constraints;
    this.threat_scenarios = args.threat_scenarios;
    this.vulnerability_ids = args.vulnerability_ids;
  }
}

// Original interface kept for reference
export interface IAttackPathRequest {
  // Primary component to analyze (focal point of the analysis)
  primary_component_id: string;
  
  // Component IDs to include in the analysis
  component_ids: string[];
  
  // Entry points and targets
  entry_point_ids?: string[];
  target_ids?: string[];
  
  // Analysis parameters
  include_chains: boolean;
  max_depth?: number;
  
  // Analysis context
  assumptions?: AttackPathAssumption[];
  constraints?: AttackPathConstraint[];
  threat_scenarios?: ThreatScenario[];
  vulnerability_ids?: string[];
}

export interface AttackPathAnalysisResult {
  analysis_id: string;
  component_count: number;
  entry_points: ComponentSummary[];
  critical_targets: ComponentSummary[];
  total_paths: number;
  high_risk_paths: number;
  total_chains: number;
  high_risk_chains: number;
  created_at: string;
  scope_id: string | null;
}

/**
 * Generate attack paths based on component connections
 * 
 * @param request - The attack path analysis request parameters
 * @returns Promise with the attack path analysis result
 */
export async function generateAttackPaths(
  request: AttackPathRequest | IAttackPathRequest
): Promise<AttackPathAnalysisResult> {
  try {
    console.log('Generating attack paths with request:', request);
    
    // Validate the request has all required fields
    if (!request.primary_component_id) {
      throw new Error('Missing required field: primary_component_id');
    }
    
    if (!request.component_ids || request.component_ids.length === 0) {
      throw new Error('Missing required field: component_ids');
    }
    
    // IMPORTANT: Ensure we have target_ids
    if (!request.target_ids || request.target_ids.length === 0) {
      throw new Error('Missing required field: target_ids. The API requires at least one target component.');
    }
    
    // Ensure primary component is in the component_ids list
    if (!request.component_ids.includes(request.primary_component_id)) {
      request.component_ids.push(request.primary_component_id);
    }
    
    // Use direct fetch for most reliable results
    const url = 'http://127.0.0.1:8080/api/attack-paths';
    console.log('Making direct API call to:', url);
    console.log('Request payload:', JSON.stringify(request, null, 2));
    
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json'
      },
      body: JSON.stringify(request)
    });
    
    if (!response.ok) {
      const errorText = await response.text();
      console.error('Error response from API:', errorText);
      throw new Error(`API error: ${response.status} - ${errorText}`);
    }
    
    // Check if the response has content
    const contentType = response.headers.get('content-type');
    if (!contentType || !contentType.includes('application/json')) {
      const text = await response.text();
      console.log('Non-JSON response:', text);
      throw new Error('API did not return JSON. Response: ' + text);
    }
    
    const data = await response.json();
    console.log('API response data:', data);
    
    if (!data) {
      throw new Error('API returned empty response');
    }
    
    return data;
  } catch (error) {
    console.error('Error generating attack paths:', error);
    throw error;
  }
}

/**
 * Get all attack paths with optional filtering
 * 
 * @param analysisId - Optional analysis ID to filter by
 * @param skip - Number of items to skip for pagination
 * @param limit - Maximum number of items to return
 * @returns Promise with the list of attack paths and total count
 */
export async function getAttackPaths(
  analysisId?: string,
  skip: number = 0,
  limit: number = 100
): Promise<AttackPathList> {
  try {
    // Build the URL exactly like the curl command that works
    let url = 'http://127.0.0.1:8080/api/attack-paths?skip=' + skip + '&limit=' + limit;
    
    if (analysisId) {
      url += '&analysis_id=' + analysisId;
    }
    
    console.log('Fetching attack paths from: ' + url);
    
    // Use simple fetch with the exact same parameters as the curl command
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'accept': 'application/json'
      }
    });
    
    if (!response.ok) {
      throw new Error('API error: ' + response.status);
    }
    
    // Get the raw data first so we can log it
    const data = await response.json();
    console.log('Raw API response:', data);
    
    // Handle what we get from the API
    if (Array.isArray(data)) {
      // If we got an array, wrap it in the expected format
      return { paths: data, total: data.length };
    } else if (data && data.paths && Array.isArray(data.paths)) {
      // If we got the expected format, return it as is
      return data;
    } else {
      // If we can't find paths, return empty array
      console.error('Unexpected API response format:', data);
      return { paths: [], total: 0 };
    }
  } catch (error) {
    console.error('Error fetching attack paths:', error);
    return { paths: [], total: 0 };
  }
}

/**
 * Get a specific attack path by ID
 * 
 * @param pathId - The attack path ID
 * @returns Promise with the attack path details
 */
export async function getAttackPath(pathId: string): Promise<AttackPath> {
  try {
    const response = await apiClient.get<AttackPath>(
      `/attack-paths/${pathId}`
    );
    return response.data;
  } catch (error) {
    console.error(`Error fetching attack path ${pathId}:`, error);
    throw error;
  }
}

/**
 * Get all attack chains with optional filtering
 * 
 * @param analysisId - Optional analysis ID to filter by
 * @param skip - Number of items to skip for pagination
 * @param limit - Maximum number of items to return
 * @returns Promise with the list of attack chains and total count
 */
export async function getAttackChains(
  analysisId?: string,
  skip: number = 0,
  limit: number = 100
): Promise<AttackChainList> {
  try {
    // Build the URL exactly like the working endpoint
    let url = 'http://127.0.0.1:8080/api/attack-paths/chains?skip=' + skip + '&limit=' + limit;
    
    if (analysisId) {
      url += '&analysis_id=' + analysisId;
    }
    
    console.log('Fetching attack chains from: ' + url);
    
    // Use simple fetch
    const response = await fetch(url, {
      method: 'GET',
      headers: {
        'accept': 'application/json'
      }
    });
    
    if (!response.ok) {
      throw new Error('API error: ' + response.status);
    }
    
    // Get the raw data first so we can log it
    const data = await response.json();
    console.log('Raw chains API response:', data);
    
    // Handle what we get from the API
    if (Array.isArray(data)) {
      // If we got an array, wrap it in the expected format
      return { chains: data, total: data.length };
    } else if (data && data.chains && Array.isArray(data.chains)) {
      // If we got the expected format, return it as is
      return data;
    } else {
      // If we can't find chains, return empty array
      console.error('Unexpected API response format for chains:', data);
      return { chains: [], total: 0 };
    }
  } catch (error) {
    console.error('Error fetching attack chains:', error);
    return { chains: [], total: 0 };
  }
}

/**
 * Get a specific attack chain by ID
 * 
 * @param chainId - The attack chain ID
 * @returns Promise with the attack chain details
 */
export async function getAttackChain(chainId: string): Promise<AttackChain> {
  try {
    const response = await apiClient.get<AttackChain>(
      `/attack-paths/chains/${chainId}`
    );
    return response.data;
  } catch (error) {
    console.error(`Error fetching attack chain ${chainId}:`, error);
    throw error;
  }
}

File: api/models/attack_path.py
"""
Attack Path Analysis models for FastAPI
"""
from enum import Enum
from typing import List, Dict, Optional, Any
from datetime import datetime
from pydantic import BaseModel, Field


class AttackPathType(str, Enum):
    """Types of attack paths"""
    DIRECT = "Direct"             # Direct path from entry point to target
    MULTI_STEP = "Multi-Step"     # Multi-step attack requiring multiple components
    LATERAL = "Lateral"           # Lateral movement across trust boundaries
    PRIVILEGE_ESCALATION = "Privilege Escalation"  # Path that increases privileges


class AttackStepType(str, Enum):
    """Types of attack steps"""
    INITIAL_ACCESS = "Initial Access"
    EXECUTION = "Execution"
    PERSISTENCE = "Persistence"
    PRIVILEGE_ESCALATION = "Privilege Escalation"
    DEFENSE_EVASION = "Defense Evasion"
    CREDENTIAL_ACCESS = "Credential Access"
    DISCOVERY = "Discovery"
    LATERAL_MOVEMENT = "Lateral Movement"
    COLLECTION = "Collection"
    EXFILTRATION = "Exfiltration"
    COMMAND_AND_CONTROL = "Command and Control"
    IMPACT = "Impact"


class AttackComplexity(str, Enum):
    """Attack complexity levels"""
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"


class StepBase(BaseModel):
    """Base model for attack steps"""
    component_id: str = Field(..., description="The component ID for this step")
    step_type: AttackStepType = Field(..., description="The type of attack step")
    description: str = Field(..., description="Description of the attack step")
    prerequisites: Optional[List[str]] = Field(default_factory=list, description="Prerequisites for this step")
    vulnerability_ids: Optional[List[str]] = Field(default_factory=list, description="Associated vulnerabilities")
    threat_ids: Optional[List[str]] = Field(default_factory=list, description="Associated threats")
    order: int = Field(..., description="The order of this step in the path")


class StepCreate(StepBase):
    """Create model for attack steps"""
    pass


class Step(StepBase):
    """Model for attack steps with ID and timestamps"""
    step_id: str = Field(..., description="Unique identifier for the step")
    created_at: datetime = Field(..., description="Creation timestamp")
    
    class Config:
        from_attributes = True


class PathBase(BaseModel):
    """Base model for attack paths"""
    name: str = Field(..., description="Name of the attack path")
    description: str = Field(..., description="Description of the attack path")
    path_type: AttackPathType = Field(..., description="Type of attack path")
    complexity: AttackComplexity = Field(..., description="Complexity of the attack path")
    entry_point_id: str = Field(..., description="Component ID of the entry point")
    target_id: str = Field(..., description="Component ID of the target")
    success_likelihood: float = Field(..., ge=0.0, le=1.0, description="Likelihood of successful exploitation")
    impact: Dict[str, int] = Field(..., description="Impact ratings (e.g., confidentiality, integrity, availability)")
    risk_score: float = Field(..., ge=0.0, le=10.0, description="Overall risk score")


class PathCreate(PathBase):
    """Create model for attack paths"""
    steps: List[StepCreate] = Field(..., description="Steps in the attack path")
    analysis_id: str = Field(..., description="ID of the analysis this path belongs to")
    scope_id: Optional[str] = Field(None, description="ID of the scope this analysis was performed in")


class Path(PathBase):
    """Model for attack paths with ID and timestamps"""
    path_id: str = Field(..., description="Unique identifier for the path")
    analysis_id: str = Field(..., description="ID of the analysis this path belongs to")
    scope_id: Optional[str] = Field(None, description="ID of the scope this analysis was performed in")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    steps: List[Step] = Field(default_factory=list, description="Steps in the attack path")
    
    class Config:
        from_attributes = True


class ChainBase(BaseModel):
    """Base model for attack chains"""
    name: str = Field(..., description="Name of the attack chain")
    description: str = Field(..., description="Description of the attack chain")
    entry_point_id: str = Field(..., description="Component ID of the primary entry point")
    final_target_id: str = Field(..., description="Component ID of the final target")
    attack_goal: str = Field(..., description="The goal of the attack")
    total_steps: int = Field(..., description="Total number of steps in the chain")
    complexity: AttackComplexity = Field(..., description="Complexity of the attack chain")
    success_likelihood: float = Field(..., ge=0.0, le=1.0, description="Likelihood of successful exploitation")
    impact: Dict[str, int] = Field(..., description="Impact ratings (e.g., confidentiality, integrity, availability)")
    risk_score: float = Field(..., ge=0.0, le=10.0, description="Overall risk score")


class ChainCreate(ChainBase):
    """Create model for attack chains"""
    path_ids: List[str] = Field(..., description="IDs of attack paths in this chain")
    analysis_id: str = Field(..., description="ID of the analysis this chain belongs to")
    scope_id: Optional[str] = Field(None, description="ID of the scope this analysis was performed in")


class Chain(ChainBase):
    """Model for attack chains with ID and timestamps"""
    chain_id: str = Field(..., description="Unique identifier for the chain")
    analysis_id: str = Field(..., description="ID of the analysis this chain belongs to")
    scope_id: Optional[str] = Field(None, description="ID of the scope this analysis was performed in")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")
    paths: List[Path] = Field(default_factory=list, description="Paths in the attack chain")
    
    class Config:
        from_attributes = True
        
    @classmethod
    def from_db_model(cls, db_model):
        """Create API model from database model"""
        return cls(
            chain_id=db_model.chain_id,
            name=db_model.name,
            description=db_model.description,
            entry_point_id=db_model.entry_points[0] if db_model.entry_points else "",
            final_target_id=db_model.targets[0] if db_model.targets else "",
            attack_goal=db_model.attack_goal,
            total_steps=sum(len(p.steps) for p in db_model.paths) if db_model.paths else 0,
            complexity=db_model.complexity,
            success_likelihood=db_model.success_likelihood,
            impact=db_model.impact,
            risk_score=db_model.risk_score,
            analysis_id=db_model.analysis_id,
            scope_id=db_model.scope_id,
            created_at=db_model.created_at,
            updated_at=db_model.updated_at,
            paths=db_model.paths
        )


class AttackPathAssumption(BaseModel):
    """Model for attack path assumptions"""
    assumption_id: str = Field(..., description="Unique identifier for this assumption")
    description: str = Field(..., description="Description of the assumption")
    type: str = Field(..., description="Type of assumption (e.g., 'physical_access', 'network_access')")


class AttackPathConstraint(BaseModel):
    """Model for attack path constraints"""
    constraint_id: str = Field(..., description="Unique identifier for this constraint")
    description: str = Field(..., description="Description of the constraint")
    type: str = Field(..., description="Type of constraint (e.g., 'exclude_physical', 'require_local')")


class ThreatScenario(BaseModel):
    """Model for predefined threat scenarios"""
    scenario_id: str = Field(..., description="Unique identifier for this threat scenario")
    name: str = Field(..., description="Name of the threat scenario")
    description: str = Field(..., description="Description of the threat scenario")
    threat_type: str = Field(..., description="Type of threat (e.g., 'spoofing', 'tampering')")
    likelihood: float = Field(..., ge=0.0, le=1.0, description="Likelihood of this threat occurring")


class AttackPathRequest(BaseModel):
    """Request model for generating attack paths"""
    # Primary component to analyze (focal point of the analysis)
    primary_component_id: str = Field(..., description="ID of the primary component to analyze")
    
    # Component IDs to include in the analysis
    component_ids: List[str] = Field(..., description="IDs of components to analyze")
    
    # Analysis metadata
    analysis_id: Optional[str] = Field(None, description="Optional existing analysis ID")
    scope_id: Optional[str] = Field(None, description="Optional scope ID")
    
    # Entry points and targets
    entry_point_ids: Optional[List[str]] = Field(None, description="Optional specific entry points")
    target_ids: Optional[List[str]] = Field(None, description="Optional specific targets")
    
    # Analysis parameters
    include_chains: bool = Field(True, description="Whether to generate attack chains")
    max_depth: Optional[int] = Field(5, description="Maximum path depth to consider")
    
    # Analysis context
    assumptions: Optional[List[AttackPathAssumption]] = Field(default_factory=list, 
                                                         description="Assumptions for this analysis")
    constraints: Optional[List[AttackPathConstraint]] = Field(default_factory=list, 
                                                          description="Constraints for this analysis")
    threat_scenarios: Optional[List[ThreatScenario]] = Field(default_factory=list, 
                                                          description="Predefined threat scenarios to consider")
    vulnerability_ids: Optional[List[str]] = Field(default_factory=list,
                                               description="IDs of vulnerabilities to consider in the analysis")


class AttackPathList(BaseModel):
    """List of attack paths"""
    paths: List[Path] = Field(..., description="List of attack paths")
    total: int = Field(..., description="Total number of paths")


class AttackChainList(BaseModel):
    """List of attack chains"""
    chains: List[Dict[str, Any]] = Field(..., description="List of attack chains")
    total: int = Field(..., description="Total number of chains")


class AttackPathAnalysisResult(BaseModel):
    """Result of an attack path analysis"""
    analysis_id: str = Field(..., description="ID of the analysis")
    component_count: int = Field(..., description="Number of components analyzed")
    entry_points: List[Dict[str, Any]] = Field(..., description="Entry points identified")
    critical_targets: List[Dict[str, Any]] = Field(..., description="Critical targets identified")
    total_paths: int = Field(..., description="Total number of attack paths identified")
    high_risk_paths: int = Field(..., description="Number of high-risk paths")
    total_chains: int = Field(..., description="Total number of attack chains identified")
    high_risk_chains: int = Field(..., description="Number of high-risk chains")
    created_at: datetime = Field(..., description="Analysis timestamp")
    scope_id: Optional[str] = Field(None, description="ID of the scope")
